File: project1/Board.rb
#Create Board object, check potential sets in the inUseCard.

require './Deck.rb'

class Board

  @deck

  def initialize
    @deck = Deck.new
  end

  #print the in used cards
  def printCurrentCards()
    @current = @deck.getCurrent()
    rowLength = @current.length / 3
    # Print Column headings
    print "     " # Extra spacing so it doesn't label the row headings with a column head
    rowLength.times { |i| print "---#{i.to_s}---" }
    puts "" #newline
    
    # Print the cards
    3.times { |i|
      print "#{(i+65).chr}:   "
      rowLength.times { |j|
        print @current[i * rowLength + j].consoleString
      }
      puts "" # Newline
    }
  end

  #method taking the card in use as an array(Deck.getCurrent)and used to determine if any set exists, if not draw three more cards
  def containSet(inUseCards)
    
    #set the indicator as the boolean expression indcating is there a set or not
    indicator = false
    
    #begin the loop to loop through all the combinations
    indexOne = 0
    while index < inUseCards.length && indicator!
      cardOne = inUseCards[indexOne]
      indexTwo = indexOne + 1
      while indexTwo < inUseCards.length && indicator!
        cardTwo = inUseCards[indexTwo]
        indexThree = indexTwo + 1
        while indexThree < inUseCards.length && indicator!
          cardThree = inUseCards[indexThree]
          indicator = checkSet(cardOne,cardTwo,cardThree)
          if (indicator!)
            indexThree += 1
          end
        end
        if (indicator!)
          indexTwo += 1
        end
      end
      if (indicator!)
        indexOne += 1
      end
    end

    if (indicator!) #if indicator shows that we do not have a set
      @deck.inUseCards.get(3)
    end
    return indicator
  end

#the function that check to see if a given combination of cards is a set
  def checkSet(a,b,c)
    indicator = true
    if (!((a.get_number == b.get_number) && (b.get_number == c.get_number) || (a.get_number != b.get_number) && (a.get_number != c.get_number) && (b.get_number != c.get_number))) 
      indicator = false
    end
    if (!((a.get_shape == b.get_shape) && (b.get_shape == c.get_shape) || (a.get_shape != b.get_shape) && (a.get_shape != c.get_shape) && (b.get_shape != c.get_shape))) 
      indicator = false
    end
    if (!((a.get_opacity == b.get_opacity) && (b.get_opacity == c.get_opacity) || (a.get_opacity != b.get_opacity) && (a.get_opacity != c.get_opacity) && (b.get_opacity != c.get_opacity))) 
      indicator = false
    end
    if (!((a.get_color == b.get_color) && (b.get_color == c.get_color) || (a.get_color != b.get_color) && (a.get_color != c.get_color) && (b.get_color != c.get_color))) 
      indicator = false
    end
    return indicator
  end

  #method used to match the string that the user entered with cards.
  def getCard (inputStr)
    #seperate three card notations using space
    cardArray = inputStr.split(' ')

    #get the col name and row number to get the index of the card in the inUseCards object
    tempOne = cardArray[0]
    cardOneCol = tempOne.first(1)
    cardOneRow = tempOne.last(1)
    tempTwo = cardArray[1]
    cardTwoCol = tempTwo.first(1)
    cardTwoRow = tempTwo.last(1)
    tempThree = cardArray[2]
    cardThreeCol = tempThree.first(1)
    cardThreeRow = tempThree.last(1)
    @cardObj = @deck.getCurrent()
    rowLength = @cardObj.length / 3

    #get three card objects and pass it to checkSet
    cardOne = @cardObj[getIndex(cardOneCol,cardOneRow,rowLength)]
    cardTwo = @cardObj[getIndex(cardTwoCol,cardTwoRow,rowLength)]
    cardThree = @cardObj[getIndex(cardThreeCol,cardThreeRow,rowLength)]
    indicator = checkSet(cardOne,cardTwo,cardThree)
    
    return indicator
  end

  #the method that take the col number and row number then return the index in the array
  def getIndex(colNum,rowNum,rowLength)
    index = 0
    if(colNum = "A")
      index = rowNum
    elsif (colNum = "B")
      index = rowLength + rowNum
    else
      index = rowLength * 2 + rowNum
    end
    
    return index
  end
  def columnCount
    return @deck.getCurrent().length / 3
  end
  def getValidSetInput
    setString = ""
    invalidInput = true
    while invalidInput do
      puts "Enter three card names separated by spaces, e.g. \"A1 B2 C3\""
      setString = gets
      setString.upcase
      cardStrings = setString.split
      cardStrings.each{ |s| 
        if ((s[0] == 'A' || s[0] == 'B' || s[0] == 'C') && (s[1].to_i <= columnCount())) then
          invalidInput = false
        end
      }
      if invalidInput then
        puts "Invalid input. Try again."
      end
    end
    return setString
  end
end

File: project2/Board.js
class Board {
    constructor() {
        this.deck = new DeckKernel();
        this.rowLength = 4;
        this.prepareBoard();
    }

    //Find some way to code analagous version of printCurrentCards(requires display of cards)
    printBoard(){
        this.prepareBoard();
    }

    // Get number of cards in board to 12, and add cards to board if there are no sets
    prepareBoard() {
        this.rowLength = 4;
        if(this.deck.inUseCards.length < 12){
            this.deck.get(12 - this.deck.inUseCards.length);
        } else if(this.deck.inUseCards.length > 12){
            this.deck.putOnTop(this.deck.inUseCards.slice(12));
        }
        if(!this.containSet()){
            this.deck.get(3);
            this.rowLength++;
        }
        if(!this.containSet()){
            this.deck.get(3);
            this.rowLength++;
        }
    }

    //method taking the card in use as an array(Deck.getCurrent)and used to determine if any set exists
    containSet() {
        let inUseCards = this.deck.inUseCards;
        //set the indicator as the boolean expression indicating is there a set or not
        let indicator = false;

        //begin the loop to loop through all combinations
        let indexOne = 0;
        while(indexOne < inUseCards.length && !indicator) {
            let cardOne = inUseCards[indexOne];
            let indexTwo = indexOne + 1;
            while(indexTwo < inUseCards.length && !indicator) {
                let cardTwo = inUseCards[indexTwo];
                let indexThree = indexTwo + 1;
                while(indexThree < inUseCards.length && !indicator) {
                    let cardThree = inUseCards[indexThree];
                    indicator = indicator || this.checkSet(cardOne, cardTwo, cardThree);
                    indexThree++;
                }
                indexTwo++;
            }
            indexOne++;
        }

        return indicator;
    }

    //the function that check to see if a given combination of cards is a set
    checkSet(a, b, c) {
        let indicator = true;
        if (!(((a.get_number == b.get_number) && (b.get_number == c.get_number))
            || ((a.get_number != b.get_number) && (a.get_number != c.get_number) && (b.get_number != c.get_number))))
          indicator = false

        if (!(((a.get_shape == b.get_shape) && (b.get_shape == c.get_shape)) 
            || ((a.get_shape != b.get_shape) && (a.get_shape != c.get_shape) && (b.get_shape != c.get_shape))))
          indicator = false

        if (!(((a.get_opacity == b.get_opacity) && (b.get_opacity == c.get_opacity))
            || ((a.get_opacity != b.get_opacity) && (a.get_opacity != c.get_opacity) && (b.get_opacity != c.get_opacity))))
          indicator = false

        if (!(((a.get_color == b.get_color) && (b.get_color == c.get_color))
            || ((a.get_color != b.get_color) && (a.get_color != c.get_color) && (b.get_color != c.get_color))))
          indicator = false

        if(a == b || a == c || b == c)
          indicator = false

        return indicator
    }

    //method used to match the string that the user entered with cards.
    getCard(inputStr) {
        let cardArray = inputStr.split(" ");

        let cardObj = this.deck.inUseCards;
        let rowLength = this.columnCount();

        let cards = [];

        for(let i = 0; i < 3; i++) {
            let temp = cardArray[i];
            cards[i] = cardObj[this.getIndex(parseInt(temp[0]), parseInt(temp[1]))];
        }

        return cards;
    }

    //the method that take the col number and row number then return the index in the array
    getIndex(rowNum, colNum) {
        return rowNum * this.rowLength + colNum;
    }

    columnCount() {
        return this.deck.inUseCards.length / 3;
    }

    getValidSetInput() {
        let setString = "";
        let validString = false;

        while(!validInput) {
            var validInput = true;
            let setString = prompt("Enter three card names seperated by spaces, e.g. \"A1 B2 C3\"");
            setString = setString.toUpperCase();
            let cardStrings = setString.split("");
            cardStrings.forEach(function(s) {
                if(!(s[0] == 'A' || s[0] == 'B' || s[0] == 'C') || !(parseInt(s[1]) < this.columnCount()))
                    validInput = false;
            });
            if(!validInput)
                console.log("Invalid input.");

            return setString;
        }
    }

    //Given 3 cards, removes them from the deck
    removeCards(array) {
        this.deck.putBack(array);
        if(this.deck.inUseCards.length < 12)
            this.deck.get(3);
    }
}
